File: python/39-combination-sum.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        n = len(candidates)
        result = []
        def dfs(i, path, target):
            if target == 0:
                result.append(path[:])
                return
            for j in range(i, n):
                candidate = candidates[j]
                if target-candidate < 0: continue
                path.append(candidate)
                dfs(j, path, target-candidate)
                path.pop()
        dfs(0, [], target)
        return result


candidates = [2,3,6,7]
target = 7
result = Solution().combinationSum(candidates, target)
print(result) # [[7], [2,2,3]]

candidates = [2,3,5]
target = 8
result = Solution().combinationSum(candidates, target)
print(result) # [[2,2,2,2], [2,3,3], [3,5]]

File: cpp/39. Combination Sum.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    vvi ans;
    void dfs(int i, int target, vi &path, vi &candidates) {
        int n = candidates.size();
        if(target == 0) {
            ans.push_back(path);
            return;
        }
        for(int j = i; j < n; j++) {
            int candidate = candidates[j];
            if(target-candidate < 0) continue;
            path.push_back(candidate);
            dfs(j, target-candidate, path, candidates);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vi path;
        dfs(0, target, path, candidates);
        return ans;
    }
};

void printAns(vvi ans) {
    for(auto l: ans) {
        cout << "(";
        for(auto n: l) {
            cout << n << " ";
        }
        cout << "), ";
    }
    cout << "\n";
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vi candidates; int target; vvi ans;

    solution = Solution();
    candidates = {2,3,6,7};
    target = 7;
    ans = solution.combinationSum(candidates, target);
    printAns(ans);

    solution = Solution();
    candidates = {2,3,5};
    target = 8;
    ans = solution.combinationSum(candidates, target);
    printAns(ans);
}